sc_cost_meter/utils.py

```diff
@@ -0,0 +1,185 @@
+  IDs to SC subscriber data
+  :param synapse_id: synapse user id
+  :return the Marketplace product code, None if cannot find customer ID
+  '''
+  product_code = None
+  ddb_marketplace_table_name = get_env_var_value('MARKETPLACE_ID_DYNAMO_TABLE_NAME')
+  if ddb_marketplace_table_name:
+    ddb_product_code_attribute = 'ProductCode'
+    client = get_dynamo_client()
+    response = client.get_item(
+      Key={
+        'SynapseUserId': {
+          'S': synapse_id,
+        }
+      },
+      TableName=ddb_marketplace_table_name,
+      ConsistentRead=True,
+      AttributesToGet=[
+        ddb_product_code_attribute
+      ]
+    )
+
+    if "Item" in response.keys():
+      product_code = response["Item"][ddb_product_code_attribute]["S"]
+      log.debug(f'marketplace product code: {product_code}')
+    else:
+      log.info(f'cannot find registration for synapse user: {synapse_id}')
+
+  return product_code
+
+
+def get_customer_cost(customer_id, time_period, granularity):
+  '''
+  Get the total cost of all resources tagged with the customer_id for a given
+  time_period.  The time_period and time granularity must match.
+  :param customer_id: the Marketplace customer ID
+  :param time_period: the cost time period
+  :param granularity: the granularity of time HOURLY|DAILY|MONTHLY
+  :return: the total cost of all resources and the currency unit
+  '''
+  client = get_ce_client()
+
+  response = client.get_cost_and_usage(
+    TimePeriod=time_period,
+    Granularity=granularity,
+    Filter={
+      "Tags": {
+        "Key": "marketplace:customerId",
+        "Values": [
+          customer_id
+        ]
+      }
+    },
+    Metrics=["UnblendedCost"]
+  )
+
+  results_by_time = response['ResultsByTime']
+  cost = results_by_time[0]["Total"]["UnblendedCost"]["Amount"]
+  unit = results_by_time[0]["Total"]["UnblendedCost"]["Unit"]
+  return cost, unit
+
+def report_cost(cost, customer_id, product_code):
+  '''
+  Report the incurred cost of the customer's resources to the AWS Marketplace
+  :param cost: the cost
+  :param customer_id: the Marketplace customer ID
+  :param product_code: the Marketplace product code
+  '''
+  cost_accrued_rate = 0.001  # TODO: use mareketplace get_dimension API to get this info
+  quantity = int(cost / cost_accrued_rate)
+
+  mrktpl_client = get_meteringmarketplace_client()
+  response = mrktpl_client.batch_meter_usage(
+    UsageRecords=[
+      {
+        'Timestamp': datetime.utcnow(),
+        'CustomerIdentifier': customer_id,
+        'Dimension': 'costs_accrued',
+        'Quantity': quantity
+      }
+    ],
+    ProductCode=product_code
+  )
+
+  status = response[0]["Status"]
+  if status == 'Success':
+    meter_record_id = response[0]["MeteringRecordId"]
+    log.info(f'recorded meter usage for customer {customer_id} '
+             f'with meter record id {meter_record_id}')
+  else:
+    # TODO: need to add a retry mechanism for failed reports
```

Comment:
I think this is just a special case of the broader question:  How do we make this robust to missing days or running more than once a day?

Reply:
yup,  will fortify after testing POC.